### PR TITLE
fix: handle ObjectGUID as Buffer correctly

### PR DESCRIPTION
**Pull request by JustSamuel**
# Description
Mixing little endian and big endian should be considered a warcrime. 
https://joelitechlife.ca/2021/04/05/get-objectguid-of-microsoft-active-directory-using-ldapsearch/

## Types of changes

- Bug fix _(non-breaking change which fixes an issue)_

### CODE REVIEW

**Comment by JustSamuel** [comment]
Note @Yoronex that this is quite a nifty migration, so this should only be done under supervision so that we can recover it if needed.

File: src/service/authentication-service.ts [hunk 1]
```diff
@@ -268,9 +268,10 @@ export default class AuthenticationService extends WithManager {
       const { searchEntries } = await client.search(ldapSettings.base, {
         scope: 'sub',
         filter: filterstr,
+        explicitBufferAttributes: ['objectGUID'],
       });
       if (searchEntries[0]) {
-        ADUser = userFromLDAP(searchEntries[0]);
+        ADUser = userFromLDAP(searchEntries[0] as any as LDAPResult);
```
Comment (Yoronex) [inline src/service/authentication-service.ts:274]: Same here.

File: test/unit/service/ad-service.ts [hunk 2]
```diff
@@ -308,6 +317,10 @@ describe('AD Service', (): void => {
         { where: { authenticateAs: { id: newOrgan.id } }, relations: ['user'] },
       )).map((mAuth) => mAuth.user.id);
 
+      console.error((await MemberAuthenticator.find(
+        { where: { authenticateAs: { id: newOrgan.id } }, relations: ['user'] },
+      )));
+      console.error(secondMembers);
```
Comment (Yoronex) [inline test/unit/service/ad-service.ts:323]: `console.error`

File: src/service/authentication-service.ts (hunk 1 above)
Comment (JustSamuel) [reply src/service/authentication-service.ts:274]: This is directly from the `client.search`, so no.

File: src/service/ad-service.ts [hunk 3]
```diff
@@ -112,7 +114,7 @@ export default class ADService extends WithManager {
       // Extract members
       const shared = sharedAccounts[i];
       const result = await this.getLDAPGroupMembers(client, shared.dn);
-      const members: LDAPUser[] = result.searchEntries.map((u) => userFromLDAP(u));
+      const members: LDAPUser[] = result.searchEntries.map((u) => userFromLDAP(u as any as LDAPResult));
```
[the remark this answers is not in the document]
Comment (JustSamuel) [reply src/service/ad-service.ts]: fixed

File: src/helpers/ad.ts [hunk 4]
```diff
@@ -67,11 +67,11 @@ export interface LDAPUser {
...
-  ADUser: { objectGUID: string }, user: User): Promise<LDAPAuthenticator> {
-  const auth = Object.assign(new LDAPAuthenticator(), {
-    user,
+  ADUser: { objectGUID: Buffer }, user: User): Promise<LDAPAuthenticator> {
+  const auth = manager.create(LDAPAuthenticator, {
+    user: user,
     UUID: ADUser.objectGUID,
-  }) as LDAPAuthenticator;
+  });
   await manager.save(auth);
```
[the remark this answers is not in the document]
Comment (JustSamuel) [reply src/helpers/ad.ts]: fixed

File: test/unit/service/ad-service.ts (hunk 2 above)
Comment (JustSamuel) [reply test/unit/service/ad-service.ts]: fixed

File: src/service/authentication-service.ts (hunk 1 above)
Comment (Yoronex) [reply src/service/authentication-service.ts:274]: Then why the need to cast it, if `searchEntries[0]` is already of type `any`?

Comment (JustSamuel) [reply src/service/authentication-service.ts:274]: It's not, its of type Entry which is an object of key-value pairs with strings as key. They decided not to specify any of the keys, not even the default AD ones which we use.

Comment (JustSamuel) [reply src/service/authentication-service.ts:274]: A possible refactor would be to have a helper function which turns the Entry into an LDAPResult and even type checks it for sanity, however I do think this is out-of-scope of this PR. Previously before this PR the any to type conversion just happened in the `userFromLDAP` function itself, so nothing changed.

**Review by Yoronex — APPROVED** [review]

**Pull request merged**